Fix sampleTerrain when using ArcGIS Terrain

DanielLeone opened the pull request:
Hi all,

I've got a WIP fix an issue and I'm keen for some feedback/suggestions.
This PR fixes #8007 

There's 2 bugs as mentioned in the issue.

1. The `availability` property was not exposed from the `ArcGISTiledElevationTerrainProvider`.
    * I've fixed that in the same way the @OmarShehata suggested - buy exposing `_tilesAvailable` 

2. calling `interpolateHeight` on a `HeightmapTerrainData` instance from `ArcGISTiledElevationTerrainProvider` will return undefined.
    * The reason this is happening is because:
        1. the data returned from an ArcGIS server is encoded in [LERC](https://www.osgeo.org/projects/lerc-limited-error-raster-compression/) format
        2. by default the `interpolateHeight` method of `HeightmapTerrainData` will try to use that array buffer (returned from the server) directly like a 2D height map array lookup thing.
        3.  this just does not work for the buffer since it's encoded in a different format
    * However we can avoid this code path by just calling `createMesh()` before calling `interpolateHeight`
        * We can see the first line of `createVerticiesFromHeightmap` (which is what `createMesh` calls) is where the LERC decoding happens.
![image](https://user-images.githubusercontent.com/8600684/102173500-16b7cd80-3ed6-11eb-8ee4-71a63abc4442.png)
        * This then forces us to use the mesh for `interpolateHeight` rather than the LERC buffer. This code path works as expected.
![image](https://user-images.githubusercontent.com/8600684/102173615-58487880-3ed6-11eb-981d-a77d9a71a230.png)

I've opted for this solution (just calling `createMesh` on every tile) inside `sampleTerrain` as it seemed to be the lowest friction change to me; but I don't think it's an optimal solution.

Another maybe better idea:
* Make the ArcGIS Terrain Provider return a different `HeightmapTerrainData` class which already has it's mesh created beforehand; and then to meet the TerrainData interface `createMesh` would just be a no-op; and `upsample` seems to use the mesh anyway to build a new height map buffer. So `MeshedHeightmapTerrainData.upsample -> HeightmapTerrainData` or something.

Current issue with this solution:
* all the `sampleTerrain` unit tests are failing and I'm not sure why. It appears when using Cesium World Terrain (as in the unit tests); `interpolateHeight` via the mesh code path isn't working for some reason 😢 . Any ideas on this one? I couldn't find any unit tests covering `QuantizedMeshTerrainData.interpolateHeight` when using the mesh code; nor do I see where it's ever really run in production (until now in this PR)... any help here is appreciated.

A less-than-ideal solution to this new issue would be
* Only call `createMesh` for ArcGIS tiles.

Other changes:
* Fixed the doc for `TerrainData.createMesh` (it was missing the terrain exaggeration param)
* Added the IntelliJ shelf to the git ignore list
* Updated the Terrain sandcastle so you can left and right click for `sampleTerrain` and `sampleTerrainMostDetailed` respectively and a few other changes - this is just temporary while testing though.


Any help is much appreciated; thanks!

cesium-concierge commented:
Thank you so much for the pull request @DanielLeone! I noticed this is your first pull request and I wanted to say welcome to the Cesium community!

The [Pull Request Guidelines](https://github.com/CesiumGS/cesium/blob/master/CONTRIBUTING.md#pull-request-guidelines) is a handy reference for making sure your PR gets accepted quickly, so make sure to skim that.

* :x: Missing `CONTRIBUTORS.md` entry.
    * Please add yourself to the [contributors](https://github.com/DanielLeone/cesium/blob/fix-arcgis-sample-terrain-height/CONTRIBUTORS.md) file!
* :heavy_check_mark: Signed CLA found.
* :grey_question: [CHANGES.md](https://github.com/CesiumGS/cesium/blob/fix-arcgis-sample-terrain-height/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.
* :grey_question: Unit tests were not updated.
    * Make sure you've [updated tests](https://github.com/CesiumGS/cesium/tree/master/Documentation/Contributors/TestingGuide) to reflect your changes, added tests for any new code, and ran the code coverage tool.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

DanielLeone commented on Source/Core/ArcGISTiledElevationTerrainProvider.js:
```diff
@@ -650,6 +657,7 @@ function requestAvailability(that, level, x, y) {
 
     // Mark whole area as having availability loaded
     that._tilesAvailablityLoaded.addAvailableTileRange(
+      level,
```
I just saw this - it looked like a missing parameter - I've got no tests or further context on this one! 😛  Will revert later.

IanLilleyT replied:
😮

IanLilleyT replied:
this makes me question `_tilesAvailablityLoaded` even more...

DanielLeone commented on Source/Core/sampleTerrain.js:
```diff
@@ -89,6 +89,10 @@ function doSampling(terrainProvider, level, positions) {
+      // Make sure to generate our mesh for each tile first
+      //  because some tiles actually require the mesh to interpolate heights
+      //  (eg: ArcGISTiledElevationTerrainProvider)
+      .then(createMeshCreatorFunction(tileRequest))
```
This is essentially the fix I've added. Call `createMesh` on each tile as it comes back. However; as mentioned, this causes CWT terrain `interpolateHeight` to return undefined  - I haven't figured out why yet :)

DanielLeone commented on Source/Core/HeightmapTerrainData.js:
```diff
@@ -182,7 +182,7 @@ Object.defineProperties(HeightmapTerrainData.prototype, {
 });
 
-var taskProcessor = new TaskProcessor("createVerticesFromHeightmap");
+var taskProcessor = new TaskProcessor("createVerticesFromHeightmap", 5000);
```
Ahhh this previous limit of 5 was interfering with tests or the sand castle or something... I can't really remember 😛  I guess because every tile now has to go across the worker (during `createMesh`), there's a lot more traffic, and therefore chances for that limit of 5 to be reached pretty quick. Will revert / figure it out later.

DanielLeone commented on Source/Core/ArcGISTiledElevationTerrainProvider.js:
```diff
@@ -304,7 +304,14 @@ Object.defineProperties(ArcGISTiledElevationTerrainProvider.prototype, {
+        );
+      }
+      //>>includeEnd('debug');
+      return this._tilesAvailable;
```
This was the other fix mentioned... I'm not really across how the `TileAvailability` class works and if this `_tilesAvailable ` property is setup correctly. I did notice there's actually two properties `_tilesAvailablityLoaded` and `_tilesAvailable` in the ArcGIS Terrain Provider; not sure which one is correct in this case.

Either way, I think this makes `sampleTerrainMostDetailed` work as expected 👍

IanLilleyT replied:
yeah, this seems like the right thing to return. I'm still wrapping my head around `_tilesAvailablityLoaded`. I think it only says whether availability has been loaded for tiles, not whether a tile is actually available. Not sure why `_tilesAvailable` can't do take care of this, but there's probably a good reason.

DanielLeone commented:
@IanLilleyT You might be interested in this one as #8481 is next on the chopping block 😄

IanLilleyT commented:
I looked over the code and the PR write-up helped get me up to speed quickly. I came to a pretty similar conclusion to you, and here's my summary (just to make sure I have all the details right): `interpolateHeights` is supposed to work for a `TerrainData` even if its triangle mesh hasn't been created yet, and it does this by accessing heights directly from the buffer that came from the server. But in this case the buffer from `ArcGisTiledElevationTerrainProvider` is LERC-encoded so it doesn't work until it has been decoded. And the current workaround is to call the `createMesh` function from inside `sampleTerrain` (which triggers the LERC-decode) in order to turn the heightmap into a mesh so it can be interpolated. I agree it's not the optimal solution because it adds some messiness to `sampleTerrain`.

Your idea of creating a new `MeshedHeightmapTerrainData` that forces the heightmap to be a mesh from the start feels like the right direction. But I think instead of creating a new `TerrainData` type, it could make sense to call the LERC-decode from inside `ArcGisTiledElevationTerrainProvider`'s `requestTileGeometry` and then create a regular `HeightmapTerrainData` from the decoded buffer. The decode call would probably need to go in a worker otherwise the main thread will be stalled with all these LERC decodes.

I'm tempted to remove custom encodings from `HeightmapTerrainData` and `createVerticesFromHeightmap` entirely, but there could be custom `TerrainProviders` out there that use them, so it would be a breaking change. But the fact that `TerrainData` functions are supposed to work pre and post `createMesh` makes me think there is a problem with the `HeightmapTerrainData` API allowing these encodings.

IanLilleyT commented:
I was thinking about this a little more and started to change my mind from my previous comment. If a `TerrainData` is supposed to be a black box that converts arbitrary terrain formats to a `TerrainMesh`, then starting out in LERC format and decoding inside `HeightmapTerrainData` makes sense. So my earlier idea of moving the LERC decode into `ArcGISTiledElevationTerrainProvider` is probably not the right move.

So I'm liking the original approach of calling `createMesh` from `sampleTerrain`. But instead of having a special check for `ArcGISTiledElevationTerrainProvider` in `sampleTerrain`, what if we make `interpolateHeight` return `undefined` when the `TerrainData` is not ready to be interpolated. Then `sampleTerrain` would call `createMesh` followed by another `interpolateHeight`. And the nice part is it's not much code change from the original PR.

DanielLeone commented:
@IanLilleyT Sounds good to me. Code has been updated! 👍 Let me know and I'll revert some of those other temporary changes.

DanielLeone commented:
Okay, one more change. I reverted the 5000 limit on the task processor - so we're back to 5 as in master.
This broke sampling if more than 5 meshes needed to be sampled.
Fixed by putting in a tasty recursive promise ⚡

IanLilleyT commented:
Nice

My two issues right now are:
1. I don't feel so good about the `sleep`. It's not really something the rest of the codebase does. We might be able to add a new option to `createMesh` like `noThrottle` to avoid needing an infinite loop in `sampleTerrain`. I hope to work on a PR for something like that today.
2. The new test data is too big. Is it possible to make it a lot smaller?

Anyway, feel free to start removing any of the temp code. Once my PR is merged we can do code review. Thanks!

DanielLeone commented:
Okay no worries. Please let me know when the `noThrottle` option is in and I'll update all these things together. 👍

IanLilleyT commented:
@DanielLeone, I opened https://github.com/CesiumGS/cesium/pull/9313. Once you have a chance to merge with that branch please set this PR's base branch to `terrainDataThrottleControl`.

Let me know if you have any questions!

DanielLeone commented:
@IanLilleyT Done!

I got it down to `180KB` of test data; I hope that's okay :)
90% of that size is the terrain tiles themselves - I've put in 1 CWT tile and 1 ArcGIS terrain tile.
I'd like to use these same tiles as part of the tests for my next PR.
Let me know if there's anything else :) 🌮

IanLilleyT commented on Specs/Data/Terrain/assets_cesium_com/layer.json:
```diff
@@ -0,0 +1,28 @@
+{
```
Cesium World Terrain data should go in `Specs/Data/CesiumTerrainTileJson` and ArcGIS data should go in `Specs/Data/ArcGIS`. I'm having trouble coming up with a naming scheme to differentiate from other files in those folders, so how about a subfolder with the tile coords as the name (e.g. `9_214_379/`)

IanLilleyT reviewed (changes requested):
@DanielLeone Overall this looks really good. In the end not a lot of code change which I love to see! I tried it out with the `Terrain.html` you had in a previous commit and everything is working as expected. There's several little comments, mostly for the specs which need to be ES5 instead of ES6.

IanLilleyT commented on Specs/Core/sampleTerrainSpec.js:
```diff
@@ -99,4 +106,183 @@ describe("Core/sampleTerrain", function () {
+
+    it("should work for Cesium World Terrain", () => {
+      patchXHRLoad({
+        "/fake/url/layer.json": "Data/Terrain/assets_cesium_com/layer.json",
```
I mainly ask this because I was trying to compare this code with existing specs by doing a global search for `fake/url` and no results showed up

DanielLeone commented:
@IanLilleyT thanks for all the suggestions 👍 I think I got them all 🌮

IanLilleyT commented on Specs/Core/sampleTerrainSpec.js:
```diff
@@ -171,43 +171,49 @@ describe("Core/sampleTerrain", function () {
-          sourceUrl.pathname + sourceUrl.search + sourceUrl.hash;
+        // find a key (source path) path in the spec which matches (ends with) the requested url
+        var availablePaths = Object.keys(proxySpec);
+        var proxiedUrl = null;
```
Do `var proxiedUrl;` and down below `if (!defined(proxiedUrl)) {` (will need to include `Source/Core/defined.js`). That's the more standard way of handling null/undefined in the repo.

IanLilleyT commented:
@DanielLeone I posted one last code request 

If all goes well https://github.com/CesiumGS/cesium/pull/9313 and this will be in the February release!

DanielLeone commented:
@IanLilleyT Thanks for writing the throttling code and the quick PR reviews! 👍 🌮

IanLilleyT commented:
Unfortunately didn't get https://github.com/CesiumGS/cesium/pull/9313 reviewed in time for the February release, so it will be out in the March release.

IanLilleyT commented:
Merged at last! Thanks @DanielLeone!

IanLilleyT commented:
Now time to peek at https://github.com/CesiumGS/cesium/pull/9389
